File: frontend/types/walkforward.ts
export type WalkForwardListItem = {
  id: number;
  created_at: string;
  strategy_name: string;
  symbol: string;
  timeframe: string;
  mode: string;
  status: string;
  robustness_score: number;
  wfe: number;
  consistency_score: number;
  average_sharpe: number;
  average_drawdown: number;
  deployment_decision: string;
};

export type WalkForwardDetail = {
  id: number;
  strategy_name: string;
  symbol: string;
  timeframe: string;
  mode: string;
  status: string;
  parameters: Record<string, unknown>;
  aggregated_metrics: Record<string, number | string>;
  combined_equity_curve: Array<{ timestamp: string; equity: number }>;
  monte_carlo_results: Record<string, number>;
  deployment_decision: { decision?: string; approved?: boolean; checks?: Record<string, boolean> };
  overfit_warnings: Array<Record<string, unknown>>;
  report: { charts?: Record<string, string> };
  windows: Array<{
    id: number;
    window_id: number;
    train_start: string;
    train_end: string;
    test_start: string;
    test_end: string;
    best_params: Record<string, unknown>;
    train_metrics: Record<string, number>;
    test_metrics: Record<string, number>;
    equity_curve: Array<{ timestamp: string; equity: number }>;
    trades: Array<Record<string, number | string>>;
    wfe: number;
    overfit_warning: boolean;
  }>;
};

File: backend/app/schemas/walkforward.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class WalkForwardStart(BaseModel):
    strategy_name: str = "momentum_breakout_v1"
    symbol: str
    timeframe: Literal["1m", "5m", "15m", "1h", "4h", "1d"] = "1h"
    mode: Literal["rolling", "expanding"] = "rolling"
    start_at: datetime
    end_at: datetime
    train_period_days: int = 365
    test_period_days: int = 90
    step_days: int = 90
    n_trials: int = Field(default=30, ge=1, le=500)
    initial_cash: Decimal = Decimal("10000")
    data_source: Literal["cache", "gateio", "csv"] = "cache"
    csv_data: str | None = None
    parameters: dict[str, Any] = Field(default_factory=dict)


class WalkForwardListItem(BaseModel):
    id: int
    created_at: datetime
    strategy_name: str
    symbol: str
    timeframe: str
    mode: str
    status: str
    robustness_score: float = 0
    wfe: float = 0
    consistency_score: float = 0
    average_sharpe: float = 0
    average_drawdown: float = 0
    deployment_decision: str = "AUTO_DEPLOYMENT_REJECT"


class WalkForwardDetail(BaseModel):
    id: int
    strategy_name: str
    symbol: str
    timeframe: str
    mode: str
    status: str
    parameters: dict[str, Any]
    aggregated_metrics: dict[str, Any]
    combined_equity_curve: list[dict[str, Any]]
    monte_carlo_results: dict[str, Any]
    deployment_decision: dict[str, Any]
    overfit_warnings: list[Any]
    report: dict[str, Any]
    windows: list[dict[str, Any]]
